apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDb.cs

```diff
@@ -83,6 +84,61 @@ public abstract class DatastoreDb
         public static DatastoreDb Create(string projectId, string namespaceId = "", DatastoreClient client = null) =>
             new DatastoreDbImpl(projectId, namespaceId, client ?? DatastoreClient.Create());
 
+        /// <summary>
+        /// Creates a <see cref="DatastoreDb"/> by either connecting to the production servers using the
+        /// default application credentials, or by connecting to a local emulator. The behavior depends on
+        /// the DATASTORE_EMULATOR_HOST and DATASTORE_PROJECT_ID environment variables, with further control
+        /// provided by the <paramref name="environmentBehavior"/> parameter.
+        /// </summary>
+        /// <param name="environmentBehavior">The way to behave in different environments. See the enum
+        /// definition for more details.</param>
+        /// <param name="projectId">The project ID to use when connecting to production servers. Ignored
+        /// when connecting to the emulator. Must not be null for production scenarios.</param>
+        /// <param name="namespaceId">The namespace ID to use in operations requiring a partition. Must not be null.</param>
+        /// <exception cref="InvalidOperationException">The environment was not configured appropriately
+        /// for the requested behavior.</exception>
+        /// <returns>A <see cref="DatastoreDb"/> operating on the specified partition, either connected
+        /// to production servers or the Datastore emulator.</returns>
+        public static DatastoreDb CreateWithEnvironment(EnvironmentBehavior environmentBehavior, string projectId, string namespaceId = "")
```

Comment:
I suspect that'll be tricky if you want your tests and your production code to go through the same path though, with only data changing. Imagine you have a config file which has the enum value in, for example... or a config class with the enum as another property. How would you use that in the EnvironmentSettings approach?

I suspect there are possibilities here though... I'll create another PR with this to see how it looks.

Reply:
But then they've got to have different actual code, rather than just data. At that point, we could just have four different methods. The intention here was to allow the *data* to drive what's going on - so all you'd need is different configuration files saying which mode to use, for example. No conditional code required at all.

What kind of error are you trying to avoid using the builder pattern here?

Reply:
In the current state developer needs to understand how the correlation of _environmentBehavior_ enum with the environment variables (or their omission) will change the behavior of the app. So it's like analyzing the state machine to decide whether given combination does what developer expects. This is not intuitive DevX. 
Having two separate methods will simplify things (at the expense of having two separate code paths on two environments).
Having the same method with **EnvironmentSettingsBuilder** will simplify things as developer can declaratively decide what to use as opposed to relying on a state machine that they need to understand and analyze. The downside here is that there may be two different code paths to create the **EnvironmentSettings**, but the actual library call from developer's standpoint is the same.

Reply:
I take issue with "it's like analyzing the state machine" - it's reading a set of descriptions which can be reasonably clearly described and named. I think that's likely to be just as intuitive as - or more than - the environment builder, to be honest.

You say "The downside here is that there may be two different code paths to create the EnvironmentSettings" - but that's exactly the problem with having two different methods to start with. You're not eliminating that, you're just moving it slightly, and I doubt that it'll be moving it enough. I'd really like there to be a solution that can use the same code path for all desired behaviors, driven only by data.

Of course we could provide *both* options potentially, but I don't *think* we want to do that.

Reply:
I agree - having both options will only confuse the developer. It should be one way that is simple to use.

As to which one - with the builder or this PR implementation - both will work, but I think builder pattern is more intuitive for developers who are used to certain development style (.NET Framework BCL), where you have connection string builders, string builders, configuration builders, etc. 
With builders the power and responsibility is in user's hands. With state machine, the power is in library's hands but the responsibility is still in user's hands.
I think that constructing configuration in two ways for two environments is acceptable as long as the resulting API call is the same.

Reply:
I'm still not convinced, but I'll create another PR to see what it looks like in a more concrete form. (I also still think referring to a state machine is over the top when discussing this. It's just method parameters, one of which is an enum. No state to keep track of. Arguably the builder approach is *more* of a state machine.)